src/inference/hmc.jl

```diff
@@ -100,21 +100,52 @@ function HMC{AD}(
     return HMC{AD}(ϵ, n_leapfrog, metricT, space)
 end
 
+function update_hamiltonian!(spl, model, n)
+    metric = gen_metric(n, spl)
+    ℓπ = gen_logπ(spl.state.vi, spl, model)
+    ∂ℓπ∂θ = gen_∂logπ∂θ(spl.state.vi, spl, model)    
+    spl.state.h = AHMC.Hamiltonian(metric, ℓπ, ∂ℓπ∂θ)
+    return spl
+end
+
 function AbstractMCMC.sample_init!(
     rng::AbstractRNG,
     model::AbstractModel,
     spl::Sampler{<:Hamiltonian},
     N::Integer;
     verbose::Bool=true,
     resume_from=nothing,
+    init_theta=nothing,
     kwargs...
 )
-
     # Resume the sampler.
     set_resume!(spl; resume_from=resume_from, kwargs...)
 
     # Get `init_theta`
     initialize_parameters!(spl; verbose=verbose, kwargs...)
+    if init_theta !== nothing
+        # Doesn't support dynamic models
+        link!(spl.state.vi, spl)
+        model(spl.state.vi, spl)
+        theta = spl.state.vi[spl]
+        spl.state.z.θ .= theta
+    else
+        # Samples new values and sets trans to true, then computes the logp
+        model(empty!(spl.state.vi), SampleFromUniform())
```

Comment:
What does this line actually do? I'm confused because it follows by the `link!`call, but I would expect the unifromlly initalized values are already in the unconstrained space.

Reply:
It puts values in `vi` which are sampled from the uniform distribution then transformed with `invlink`. This happens in `inittrans` in https://github.com/TuringLang/DynamicPPL.jl/blob/master/src/utils.jl#L131.

Reply:
Looking at the model I don't update the phase point which I should to use the transformed logp and gradient. I will fix that. @yebai please don't release yet.

Reply:
>  I don't update the phase point which I should to use the transformed logp and gradient

I believe we had some similar discussion before when we visit this issue for the first time. IRRC, we had figure out a way to set the initial value in the very beginning s.t. we don't need to manually make the phase point updated, which is a bit unintuitive to ensure everywhere outside AHMC.

Reply:
I think we do to compute the logp and its gradient at the initial theta in the unconstrained space and check that it is finite.

Reply:
This is done while constructing the phase point in AHMC.

What I mean by my comment above is that, i think the correctly logic in Turing is the following

1. Do whatever initialization for a constrained value
2. Link it to unconstrained value
3. Initial a phase point
4. Rest...

If the order is so, there is no need to update a phase point.

Reply:
(surely we still need to do this in Gibbs as you pointed in the new PR)